### PR TITLE
[cdac] Implement GetThreadStoreData in cDAC

### PR DESCRIPTION
**Pull request by elinor-fung**
- Implement `GetThreadStoreData` and `GetThreadCounts` in `Thread` contract
- Finish implementing `ISOSDacInterface::GetThreadStoreData` in cDAC
  - Add specific threads (first in thread store, Finalizer, GC) and counts
- Make existing DAC call into cDAC for `GetThreadData` if available
  - Only fills out managed thread ID and next thread right now - always returns E_NOTIMPL
- Update the example C# API in docs to be closer to what we have now

Contributes to https://github.com/dotnet/runtime/issues/99298

cc @lambdageek

### CODE REVIEW

File: docs/design/datacontracts/contract_csharp_api_design.cs [hunk 1]
```diff
@@ -72,109 +72,53 @@ struct FieldLayout
...
-    }
-
     interface IContract
     {
         string Name { get; }
         uint Version { get; }
     }
-        class Target
+
+    class Target
```
Comment (AaronRobinsonMSFT) [inline docs/design/datacontracts/contract_csharp_api_design.cs:81]: Can this be `sealed`?

File: src/coreclr/debug/runtimeinfo/datadescriptor.h [hunk 2]
```diff
@@ -105,14 +105,21 @@ CDAC_TYPES_BEGIN()
 
 CDAC_TYPE_BEGIN(Thread)
 CDAC_TYPE_INDETERMINATE(Thread)
+CDAC_TYPE_FIELD(Thread, /*uint32*/, Id, cdac_offsets<Thread>::Id)
```
Comment (AaronRobinsonMSFT) [inline src/coreclr/debug/runtimeinfo/datadescriptor.h:108]: I don't understand this comment type. Is that intentional going forward or a temporary mechanism?

File: src/native/managed/cdacreader/src/Data/Thread.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    public Thread(Target target, TargetPointer address)
+    {
+        Target.TypeInfo type = target.GetTypeInfo(DataType.Thread);
+
+        Id = target.Read<uint>(address + (ulong)type.Fields[nameof(Id)].Offset);
+        LinkNext = target.ReadPointer(address + (ulong)type.Fields[nameof(LinkNext)].Offset);
+    }
+
+    public uint Id { get; init; }
+    internal TargetPointer LinkNext { get; init; }
```
Comment (AaronRobinsonMSFT) [inline src/native/managed/cdacreader/src/Data/Thread.cs:17]: Is `internal` needed here, can this be `public` given the other members?

**Review by AaronRobinsonMSFT — COMMENTED** [review]
LGTM

File: src/native/managed/cdacreader/src/Contracts/Thread.cs [hunk 4]
```diff
@@ -35,24 +49,77 @@ static IContract IContract.Create(Target target, int version)
...
+    ThreadData IThread.GetThreadData(TargetPointer threadPointer)
+    {
+        Data.Thread? thread;
+        if (!_target.ProcessedData.TryGet(threadPointer, out thread))
+        {
+            thread = new Data.Thread(_target, threadPointer);
+
+            // Still okay if processed data is already registered by someone else.
+            _ = _target.ProcessedData.TryRegister(threadPointer, thread);
         }
```
Comment (lambdageek) [inline src/native/managed/cdacreader/src/Contracts/Thread.cs:110]: I wish we could make something reusable for this pattern.  How wasteful (I guess primarily in terms of generated code size) would passing a closure be:

```csharp
   thread = _target.ProcessedData.GetOrAdd(threadPointer, static (target, addr) => new Data.Thread(target, addr));
```

```csharp
// DataCache
public T GetOrAdd<T>(TargetPointer addr, Func<Target,TargetPointer,T> ctor) {
  if (TryGet(addr, out T result) {
    return result;
  } else {
    T constructed = ctor(_target, addr);
    if (TryRegister(addr, constructed)) {
      return constructed;
    } else {
      bool found = TryGet (addr, out T result);
      Debug.Assert(found);
      return result;
    }
  }
}
```

**Review by lambdageek — APPROVED** [review]
lgtm, although I'm not a huge fan of the TryGet TryRegister pattern.

File: src/native/managed/cdacreader/src/Contracts/Thread.cs (hunk 4 above)
Comment (lambdageek) [reply src/native/managed/cdacreader/src/Contracts/Thread.cs:110]: that's nice

**Pull request merged**